### PR TITLE
Initial commit of file support

### PR DESCRIPTION
**Pull request by DemiMarie**
This only opens and closes a handle, and does not do anything interesting.

### CODE REVIEW

**Comment by DemiMarie** [comment]
The test now manages to print out the output of `file(1)` on itself, which is interesting.

File: contrib/file/file_sapi_test.cc [hunk 1]
```diff
@@ -0,0 +1,104 @@
...
+  ASSERT_NE(fd.GetValue(), -1);
+  ASSERT_THAT(sandbox_->TransferToSandboxee(&fd), IsOk());
+  sapi::v::NullPtr null;
+  auto load_res = api_->magic_load(&magic_p, &null);
+  ASSERT_THAT(load_res, IsOk());
+  if (load_res.value()) {
+    std::cerr << GetMagicErrorStr(&magic_p) << std::endl;
+    FAIL();
+  }
+  std::cerr << "Loaded!" << std::endl;
```
Comment (cblichmann) [inline contrib/file/file_sapi_test.cc:80]: Remove these debug messages

File: contrib/file/file_sapi.h [hunk 2]
```diff
@@ -0,0 +1,59 @@
...
+        .AllowExit()
+        .AllowMmap()
+        .AllowSyscalls({
+            __NR_futex,
+            __NR_close,
+            __NR_lseek,
+            __NR_getpid,
+            __NR_clock_gettime,
+            __NR_recvmsg,
+            __NR_openat,
```
Comment (cblichmann) [inline contrib/file/file_sapi.h:42]: Use `AllowOpen()` instead. That will allow both `openat()` and plain `open()`. glibc may use either, depending on its version.

File: contrib/file/file_sapi.h [hunk 3]
```diff
@@ -0,0 +1,59 @@
...
+ public:
+  std::unique_ptr<sandbox2::Policy> ModifyPolicy(
+      sandbox2::PolicyBuilder*) override {
+    auto builder = sandbox2::PolicyBuilder()
+        .AllowSystemMalloc()
+        .AllowRead()
+        .AllowStat()
+        .AllowWrite()
+        .AllowExit()
+        .AllowMmap()
```
Comment (cblichmann) [inline contrib/file/file_sapi.h:34]: Perhaps use `AddPolicyOnMmap()` to restrict to the mapping to be read-only and no-exec?

File: contrib/file/file_sapi.h [hunk 4]
```diff
@@ -0,0 +1,59 @@
...
+            __NR_futex,
+            __NR_close,
+            __NR_lseek,
+            __NR_getpid,
+            __NR_clock_gettime,
+            __NR_recvmsg,
+            __NR_openat,
+            __NR_mprotect,
+        });
+    for (auto p : {"/etc/magic.mgc", "/usr/share/misc/magic.mgc"}) {
```
Comment (cblichmann) [inline contrib/file/file_sapi.h:45]: My system has
```
/usr/lib/file/magic.mgc
/usr/share/file/magic.mgc
/usr/share/misc/magic.mgc
```

File: contrib/file/file_sapi.h [hunk 5]
```diff
@@ -0,0 +1,59 @@
...
+      struct stat64 st;
+      if (::lstat64(p, &st) == 0 && S_ISREG(st.st_mode)) builder.AddFile(p);
```
Comment (cblichmann) [inline contrib/file/file_sapi.h:47]: Use `sapi::file_util::fileops::Exists()` here.

File: contrib/file/CMakeLists.txt [hunk 6]
```diff
@@ -0,0 +1,76 @@
...
+add_executable(file_sapi_test file_sapi_test.cc)
+
+target_link_libraries(file_sapi_test PRIVATE
+  file_sapi
+  sapi::base
+  gtest
+  gmock
+)
+
+gtest_discover_tests(file_sapi_test PROPERTIES ENVIRONMENT "TEST_FILES_DIR=${PROJECT_SOURCE_DIR}/testdata")
```
Comment (cblichmann) [inline contrib/file/CMakeLists.txt:76]: This should only be enabled if we're building the tests `SAPI_ENABLE_TESTS`

File: contrib/file/CMakeLists.txt [hunk 7]
```diff
@@ -0,0 +1,76 @@
...
+  set(SAPI_ROOT "../.." CACHE PATH "Path to the Sandboxed API source tree")
+  add_subdirectory("${SAPI_ROOT}"
+                   "${CMAKE_BINARY_DIR}/sandboxed-api-build"
+                   EXCLUDE_FROM_ALL)
+endif()
+
+find_package(PkgConfig REQUIRED)
+pkg_check_modules(S_FILE REQUIRED IMPORTED_TARGET libmagic)
+
+add_sapi_library(file_sapi
```
Comment (cblichmann) [inline contrib/file/CMakeLists.txt:38]: Please add an alias library `sapi_contrib::file`

**Review by cblichmann — CHANGES_REQUESTED** [review]

**Pull request closed without merging**